test/cosign_e2e_test.go

```diff
@@ -0,0 +1,626 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   testNamespace,
+				Labels: readNamespaceLabelsFromEnv(),
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err)
+	})
+
+	it.After(func() {
+		for tag := range builtImages {
+			deleteImageTag(t, tag)
+		}
+	})
+
+	it.Before(func() {
+		secret, err := cfg.makeRegistrySecret(dockerSecret, testNamespace)
+		require.NoError(t, err)
+
+		_, err = clients.k8sClient.CoreV1().Secrets(testNamespace).Create(ctx, secret, metav1.CreateOptions{})
+		require.NoError(t, err)
+
+		_, err = clients.k8sClient.CoreV1().ServiceAccounts(testNamespace).Create(ctx, &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: serviceAccountName,
+			},
+			Secrets: []corev1.ObjectReference{
+				{
+					Name: dockerSecret,
+				},
+			},
+			ImagePullSecrets: []corev1.LocalObjectReference{
+				{
+					Name: dockerSecret,
+				},
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err)
+
+		_, err = clients.client.KpackV1alpha2().ClusterStores().Create(ctx, &buildapi.ClusterStore{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: clusterStoreName,
+			},
+			Spec: buildapi.ClusterStoreSpec{
+				Sources: []corev1alpha1.ImageSource{
+					{Image: "gcr.io/paketo-buildpacks/bellsoft-liberica"},
+					{Image: "gcr.io/paketo-buildpacks/gradle"},
+					{Image: "gcr.io/paketo-buildpacks/syft"},
+					{Image: "gcr.io/paketo-buildpacks/executable-jar"},
+					{Image: "gcr.io/paketo-buildpacks/dist-zip"},
+					{Image: "gcr.io/paketo-buildpacks/spring-boot"},
+				},
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err)
+
+		_, err = clients.client.KpackV1alpha2().Buildpacks(testNamespace).Create(ctx, &buildapi.Buildpack{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: buildpackName,
+			},
+			Spec: buildapi.BuildpackSpec{
+				ImageSource: corev1alpha1.ImageSource{
+					Image: "gcr.io/paketo-buildpacks/bellsoft-liberica",
+				},
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err)
+
+		_, err = clients.client.KpackV1alpha2().ClusterBuildpacks().Create(ctx, &buildapi.ClusterBuildpack{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: clusterBuildpackName,
+			},
+			Spec: buildapi.ClusterBuildpackSpec{
+				ImageSource: corev1alpha1.ImageSource{
+					Image: "gcr.io/paketo-buildpacks/nodejs",
+				},
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err)
+
+		_, err = clients.client.KpackV1alpha2().ClusterStacks().Create(ctx, &buildapi.ClusterStack{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: clusterStackName,
+			},
+			Spec: buildapi.ClusterStackSpec{
+				Id: "io.buildpacks.stacks.bionic",
+				BuildImage: buildapi.ClusterStackSpecImage{
+					Image: "gcr.io/paketo-buildpacks/build:base-cnb",
+				},
+				RunImage: buildapi.ClusterStackSpecImage{
+					Image: "gcr.io/paketo-buildpacks/run:base-cnb",
+				},
+			},
+		}, metav1.CreateOptions{})
+		require.NoError(t, err)
+	})
+
+	it("Signs a Builder image successfully when the key is not password-protected", func() {
```

Comment:
Can you add 2 tests for:

- the failure case, where an invalid key (or password) should mark the builder as failed
- multiple secrets case, where a couple different secrets should result in different signatures

Reply:
@chenbh about this, I wasn't able to find a specific failure condition that gets written in the Builder/ClusterBuilder. What's the preferred method to tell a definitive failure apart from a transitory state caused by the reconcile loop in the context of kpack?

Reply:
I've decided to use the message for now, but perhaps an RFC for changing this behaviour would be warranted? It's hard to tell the failure modes apart right now.

Reply:
I think just the error message is fine for now. 

The transitive error vs definitive errors is definitely something we've run into in other parts of kpack especially in regards to network hiccups (#1091). But the whole discussion around how to detect and potentially recover from transient errors is a much larger one and should be a separate issue/pr/rfc